StyleSheets/StyleSheets/StyleSheet.swift

```diff
@@ -1,29 +1,152 @@
-import UIKit
+import Foundation
 
-public protocol StyleSheetProtocol {
-    associatedtype Element
+/// Represent a bunch of properties to be applied to the given view.
+public struct StyleSheet<View>: Equatable {
+    public typealias Inverse = StyleSheet<View>
+    
+    private var entries: [EntryKey: Any]
+    private var orderedKeys: [EntryKey]
 
-    func apply(to element: Element)
-}
+    public init() {
+        entries = [:]
+        orderedKeys = []
+    }
 
-extension StyleSheetProtocol {
     @discardableResult
-    public func compose<T>(_ property: ReferenceWritableKeyPath<Self, T>, _ value: T) -> Self {
-        self[keyPath: property] = value
-        return self
+    public func apply(to view: inout View) -> Inverse {
+        // NOTE: To handle entries with partially overlapping key paths correctly, we apply changes based on the
+        //       insertion order, and produce an inverse that applies in the reversed insertion order.
+        var inverse = StyleSheet()
+        inverse.orderedKeys = orderedKeys.reversed()
+
+        for key in orderedKeys {
+            inverse.entries[key] = key.keyPath.current(in: view)
+            key.keyPath.unsafeApply(to: &view, erasedValue: entries[key]!)
+        }
+
+        return inverse
     }
 
-    @discardableResult
-    public func compose<T>(_ property: ReferenceWritableKeyPath<Self, T?>, _ value: T?) -> Self {
-        self[keyPath: property] = value
-        return self
+    /// Retrieve the value at the specific key path.
+    public func value<Value: Equatable>(for keyPath: WritableKeyPath<View, Value>) -> Value? {
+        let key = EntryKey(keyPath)
+        return entries[key] as! Value?
+    }
+
+    public mutating func set<Value: Equatable>(_ keyPath: WritableKeyPath<View, Value>, _ newValue: Value) {
+        let key = EntryKey(keyPath)
+        entries[key] = newValue
+        orderedKeys.append(key)
```

Comment:
Probably need to consider about multiple same `keyPath` being set.

Reply:
Please add test case for this scenario as well.
(In the meeting, we conluded this as "not inverting two values with same key, but only once")

Reply:
Yeah. I was wrong in the meeting, since I forgot about having made it to record insertion order of the keys. 😅